File: frontend/src/lib/api/usage.ts
import type { ApiResponse } from './client'
import { apiGet } from './client'

// ─── Types ───

export interface UsageSummaryItem {
  event_type: string
  total_count: number
  event_count: number
}

export interface UsageSummaryResponse {
  organization_id: string | null
  period_start: string
  period_end: string
  items: UsageSummaryItem[]
  total_events: number
}

export interface UsageTimelinePoint {
  period: string
  count: number
}

export interface UsageTimelineResponse {
  organization_id: string | null
  event_type: string
  granularity: string
  points: UsageTimelinePoint[]
}

// ─── API functions ───

export function fetchUsageSummary(
  params?: { organization_id?: string; period_start?: string; period_end?: string },
  options?: { signal?: AbortSignal }
): Promise<ApiResponse<UsageSummaryResponse>> {
  const search = new URLSearchParams()

  if (params?.organization_id) search.set('organization_id', params.organization_id)
  if (params?.period_start) search.set('period_start', params.period_start)
  if (params?.period_end) search.set('period_end', params.period_end)

  const qs = search.toString()

  return apiGet<UsageSummaryResponse>(`/api/v1/usage/summary${qs ? `?${qs}` : ''}`, options)
}

export function fetchUsageTimeline(
  eventType: string,
  params?: { organization_id?: string; granularity?: string; period_start?: string; period_end?: string },
  options?: { signal?: AbortSignal }
): Promise<ApiResponse<UsageTimelineResponse>> {
  const search = new URLSearchParams({ event_type: eventType })

  if (params?.organization_id) search.set('organization_id', params.organization_id)
  if (params?.granularity) search.set('granularity', params.granularity)
  if (params?.period_start) search.set('period_start', params.period_start)
  if (params?.period_end) search.set('period_end', params.period_end)

  return apiGet<UsageTimelineResponse>(`/api/v1/usage/timeline?${search}`, options)
}

export function fetchEventTypes(
  options?: { signal?: AbortSignal }
): Promise<ApiResponse<string[]>> {
  return apiGet<string[]>('/api/v1/usage/event-types', options)
}

File: uncase/schemas/usage.py
"""API request/response schemas for usage metering."""

from __future__ import annotations

from pydantic import BaseModel, Field

EVENT_TYPES = [
    "seed_created",
    "conversation_generated",
    "evaluation_run",
    "sandbox_launched",
    "gateway_call",
    "plugin_installed",
    "knowledge_uploaded",
]


class UsageEventRecord(BaseModel):
    """Internal schema for recording a usage event."""

    organization_id: str | None = None
    event_type: str = Field(..., description="One of the supported event types")
    resource_id: str | None = None
    count: int = Field(default=1, ge=1)
    metadata: dict[str, object] | None = None
    ip_address: str | None = None


class UsageEventResponse(BaseModel):
    """A single usage event returned by the API."""

    id: str
    organization_id: str | None
    event_type: str
    resource_id: str | None
    count: int
    metadata: dict[str, object] | None
    created_at: str

    model_config = {"from_attributes": True}


class UsageSummaryItem(BaseModel):
    """Aggregated count for a single event type."""

    event_type: str
    total_count: int
    event_count: int = Field(description="Number of individual events (rows)")


class UsageSummaryResponse(BaseModel):
    """Aggregated usage summary for an organization."""

    organization_id: str | None
    period_start: str
    period_end: str
    items: list[UsageSummaryItem]
    total_events: int


class UsageTimelinePoint(BaseModel):
    """A single data point in a time-series."""

    period: str = Field(description="ISO date or datetime for the bucket")
    count: int


class UsageTimelineResponse(BaseModel):
    """Time-series usage data for charting."""

    organization_id: str | None
    event_type: str
    granularity: str
    points: list[UsageTimelinePoint]
